packages/backend-common/src/reading/tree/ZipArchiveResponse.ts

```diff
@@ -116,45 +151,41 @@ export class ZipArchiveResponse implements ReadTreeResponse {
     }
 
     const archive = archiver('zip');
-    await this.stream
-      .pipe(unzipper.Parse())
-      .on('entry', (entry: Entry) => {
-        if (entry.type === 'File' && this.shouldBeIncluded(entry)) {
-          archive.append(entry, { name: this.getInnerPath(entry.path) });
-        } else {
-          entry.autodrain();
-        }
-      })
-      .promise();
+    const tmpFile = await this.streamToTemporaryFile(this.stream);
+
+    await this.forEveryZipEntry(tmpFile, async (entry, content) => {
+      archive.append(await streamToBuffer(content), {
+        name: this.getInnerPath(entry.fileName),
+      });
+    });
+
     archive.finalize();
 
     return archive;
   }
 
   async dir(options?: ReadTreeResponseDirOptions): Promise<string> {
     this.onlyOnce();
-
     const dir =
       options?.targetDir ??
       (await fs.mkdtemp(platformPath.join(this.workDir, 'backstage-')));
 
-    await this.stream
-      .pipe(unzipper.Parse())
-      .on('entry', async (entry: Entry) => {
-        // Ignore directory entries since we handle that with the file entries
-        // as a zip can have files with directories without directory entries
-        if (entry.type === 'File' && this.shouldBeIncluded(entry)) {
-          const entryPath = this.getInnerPath(entry.path);
-          const dirname = platformPath.dirname(entryPath);
-          if (dirname) {
-            await fs.mkdirp(platformPath.join(dir, dirname));
-          }
-          entry.pipe(fs.createWriteStream(platformPath.join(dir, entryPath)));
-        } else {
-          entry.autodrain();
-        }
-      })
-      .promise();
+    const tmpFile = await this.streamToTemporaryFile(this.stream);
+
+    await this.forEveryZipEntry(tmpFile, async (entry, content) => {
+      const entryPath = this.getInnerPath(entry.fileName);
+      const dirname = platformPath.dirname(entryPath);
+
+      if (dirname) {
+        await fs.mkdirp(platformPath.join(dir, dirname));
+      }
+      return new Promise(async (resolve, reject) => {
+        const file = fs.createWriteStream(platformPath.join(dir, entryPath));
+        file.on('error', reject);
+        file.on('finish', resolve);
+        content.pipe(file);
```

Comment:
probably want to listen to those events on the pipe instead right? Thinking if `content` emits an error. Perhaps it works just fine as is but confused me a bit 😅